File: frontend/src/api/recetteFermentation.ts
import client from './client'

export interface RecetteFermentationLigne {
  id_ligne?:    number
  id_produit:   number
  quantite:     number
  unite?:       string
  note_ligne?:  string
  ordre:        number
  nom_produit?: string
  type_produit?: string
}

export interface RecetteFermentation {
  id_recette_ferm:    number
  nom_recette:        string
  type_fermentation?: string
  volume_total?:      number
  unite_volume?:      string   // L | mL
  duree_fermentation?: number  // heures
  notes?:             string
  lignes:             RecetteFermentationLigne[]
}

export type RecetteFermentationCreate = Omit<RecetteFermentation, 'id_recette_ferm'>
export type RecetteFermentationUpdate = Partial<RecetteFermentationCreate>

export const recetteFermentationAPI = {
  getAll:    ()                                                => client.get<RecetteFermentation[]>('/recettes-fermentation/'),
  getById:   (id: number)                                      => client.get<RecetteFermentation>(`/recettes-fermentation/${id}`),
  create:    (data: RecetteFermentationCreate)                 => client.post<RecetteFermentation>('/recettes-fermentation/', data),
  update:    (id: number, data: RecetteFermentationUpdate)     => client.put<RecetteFermentation>(`/recettes-fermentation/${id}`, data),
  delete:    (id: number)                                      => client.delete(`/recettes-fermentation/${id}`),
  exportCSV: ()                                                => client.get('/recettes-fermentation/export/csv', { responseType: 'blob' }),
  importCSV: (file: File) => {
    const form = new FormData(); form.append('file', file)
    return client.post('/recettes-fermentation/import', form, { headers: { 'Content-Type': 'multipart/form-data' } })
  },
}

File: backend/app/schemas/recette_fermentation.py
"""Schemas Pydantic pour RecetteFermentation"""
from pydantic import BaseModel, ConfigDict
from typing import Optional, List


class RecetteFermentationLigneBase(BaseModel):
    id_produit: int
    quantite:   float
    unite:      Optional[str] = None   # mL | L | g | Kg
    note_ligne: Optional[str] = None
    ordre:      int = 0


class RecetteFermentationLigneCreate(RecetteFermentationLigneBase):
    pass


class RecetteFermentationLigneRead(RecetteFermentationLigneBase):
    id_ligne:     int
    nom_produit:  Optional[str] = None
    type_produit: Optional[str] = None
    model_config  = ConfigDict(from_attributes=True)


class RecetteFermentationBase(BaseModel):
    nom_recette:       str
    type_fermentation: Optional[str]  = None
    volume_total:      Optional[float]= None
    unite_volume:      Optional[str]  = None   # L | mL
    duree_fermentation: Optional[int] = None   # heures
    notes:             Optional[str]  = None


class RecetteFermentationCreate(RecetteFermentationBase):
    lignes: List[RecetteFermentationLigneCreate] = []


class RecetteFermentationUpdate(BaseModel):
    nom_recette:       Optional[str]  = None
    type_fermentation: Optional[str]  = None
    volume_total:      Optional[float]= None
    unite_volume:      Optional[str]  = None
    duree_fermentation: Optional[int] = None
    notes:             Optional[str]  = None
    lignes:            Optional[List[RecetteFermentationLigneCreate]] = None


class RecetteFermentationRead(RecetteFermentationBase):
    id_recette_ferm: int
    lignes:          List[RecetteFermentationLigneRead] = []
    model_config     = ConfigDict(from_attributes=True)
